lib/Dialect/FIRRTL/Transforms/LowerGroups.cpp

```diff
@@ -0,0 +1,321 @@
+//===- LowerGroups.cpp - Lower Optiona Groups to Instances ------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// This pass lowers FIRRTL optional groups to modules and instances.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Mutex.h"
+#include "llvm/Support/RWMutex.h"
+
+#define DEBUG_TYPE "firrtl-lower-groups"
+
+using namespace circt;
+using namespace firrtl;
+
+class LowerGroupsPass : public LowerGroupsBase<LowerGroupsPass> {
+  /// Safely build a new module with a given namehint.  This handles geting a
+  /// lock to modify the top-level circuit.
+  FModuleOp buildNewModule(OpBuilder &builder, Location location,
+                           Twine namehint, SmallVectorImpl<PortInfo> &ports);
+
+  /// Function to process each module.
+  void runOnModule(FModuleOp moduleOp);
+
+  /// Entry point for the function.
+  void runOnOperation() override;
+
+  /// Indicates exclusive access to modify the circuitNamespace and the circuit.
+  llvm::sys::SmartMutex<true> *circuitMutex;
+
+  /// A circuit-level namespace.
+  CircuitNamespace *circuitNamespace;
+};
+
+/// Multi-process safe function to build a module in the circuit and return it.
+/// The name provided is only a namehint for the module---a unique name will be
+/// generated if there are conflicts with the namehint in the circuit-level
+/// namespace.
+FModuleOp LowerGroupsPass::buildNewModule(OpBuilder &builder, Location location,
+                                          Twine namehint,
+                                          SmallVectorImpl<PortInfo> &ports) {
+  llvm::sys::SmartScopedLock<true> instrumentationLock(*circuitMutex);
+  StringRef newModuleName = circuitNamespace->newName(namehint);
```

Comment:
Is the name deterministic here?

Reply:
The name is not determinstic as it depends on who gets the lock first. This is technically allowed from the FIRRTL ABI (the names of these generated modules is implementation defined). However, it's not great if names can change in the output Verilog across runs with the same source.

If we want to keep parallelism in this pass, then the names need to be either determinstic (via some mechanism that guarantees collision avoidance) or pre-allocated (where collisions are resolved serially before the parallel region).